sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventProcessorClient.java

```diff
@@ -283,13 +290,36 @@ public synchronized void start() {
      * <!-- end com.azure.messaging.eventhubs.eventprocessorclient.startstop -->
      */
     public synchronized void stop() {
+        try {
+            stop(DEFAULT_STOP_TIMEOUT);
+        } catch (RuntimeException e) {
+            logger.info("Error while stopping the event processor", e);
+        }
+    }
+
+    /**
+     * Stops processing events for all partitions owned by this event processor. All {@link PartitionProcessor} will be
+     * shutdown and any open resources will be closed.
+     *
+     * <p>
+     * Subsequent calls to stop will be ignored if the event processor is not running or is being stopped.
+     * </p>
+     *
+     * @param timeout The maximum amount of time to wait for the processor to stop processing.
+     *
+     * @throws RuntimeException if the event processor encounters timout or another error while stopping.
+     */
+    public synchronized void stop(Duration timeout) {
         if (!isRunning.compareAndSet(true, false)) {
             logger.info("Event processor has already stopped");
             return;
         }
         runner.get().cancel(true);
+
         scheduler.get().shutdown();
-        stopProcessing();
+
+        Mono.when(Mono.fromCallable(() -> scheduler.get().awaitTermination(timeout.toMillis(), TimeUnit.MILLISECONDS)), stopProcessing())
```

Comment:
I found the pattern of clean shutdown outlined as

```java
public void shutdownWithAwait(ExecutorService service) {
    service.shutdown();
    try {
        if (!service.awaitTermination(60, TimeUnit.SECONDS)) {
            service.shutdownNow();
        }
    } catch (InterruptedException ex) {
        service.shutdownNow();
        Thread.currentThread().interrupt();
    }
}
```

This was copied from [here](https://www.baeldung.com/java-executor-wait-for-threads#:~:text=When%20using%20an%20Executor%2C%20we,using%20the%20awaitTermination()%20method.)

Reply:
I believe catching and clearing the interrupt flag allows `when` operator to continue with the second publisher (i.e. stopProcessing()), avoiding pumps to stay around in case of `InterruptedException` from `awaitTermination`. Earlier with just `shutdown` there was no potential of `InterruptedException`

Reply:
thanks! updated and also added `onErrorResume` to all tasks so that if error happens other tasks will not be cancelled/dropped.